mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -589,28 +591,79 @@ def log_metric(self, run_id, metric):
         else:
             value = metric.value
-        with self.ManagedSessionMaker() as session:
-            run = self._get_run(run_uuid=run_id, session=session)
-            self._check_run_is_active(run)
-            # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
-            logged_metric, just_created = self._get_or_create(
-                model=SqlMetric,
-                run_uuid=run_id,
-                key=metric.key,
-                value=value,
-                timestamp=metric.timestamp,
-                step=metric.step,
-                session=session,
-                is_nan=is_nan,
-            )
-            # Conditionally update the ``latest_metrics`` table if the logged metric  was not
-            # already present in the ``metrics`` table. If the logged metric was already present,
-            # we assume that the ``latest_metrics`` table already accounts for its presence
-            if just_created:
-                self._update_latest_metric_if_necessary(logged_metric, session)
+        return metric, value, is_nan
 
-    @staticmethod
-    def _update_latest_metric_if_necessary(logged_metric, session):
+    def log_metric(self, run_id, metric):
+        # simply call _log_metrics and let it handle the rest
+        self._log_metrics(run_id, [metric])
+
+    def _log_metrics(self, run_id, metrics):
+        if not metrics:
+            return
+
+        # Duplicate metric values are eliminated here to maintain
+        # the same behavior in log_metric
+        metric_instances = []
+        seen = set()
+        for metric in metrics:
+            metric, value, is_nan = self._get_metric_value_details(metric)
+            if metric not in seen:
+                metric_instances.append(
+                    SqlMetric(
+                        run_uuid=run_id,
+                        key=metric.key,
+                        value=value,
+                        timestamp=metric.timestamp,
+                        step=metric.step,
+                        is_nan=is_nan,
+                    )
+                )
+            seen.add(metric)
+
+        # use SessionManager instead of ManagedSessionManager to set
+        # isolation level for individual transaction.
+        # Here, we open a connection with SERIALIZABLE isolation level.
+        # it will be released when session.commit is called.
+        session = self.SessionMaker()
+        session.connection(execution_options={"isolation_level": "SERIALIZABLE"})
+
+        run = self._get_run(run_uuid=run_id, session=session)
+        self._check_run_is_active(run)
+
+        # commit the session to make sure that we catch any IntegrityError
+        # and try to handle them.
+        try:
+            self._save_to_db(session=session, objs=metric_instances)
+            self._update_latest_metrics_if_necessary(metric_instances, session)
```

Comment:
I think that this might be the causing "deadlock", since we operate on metric_instances that are not committed yet? these are another tests that do fail on windows. it might be less performant but I would suggest something like below

```
try:
    self._save_to_db(session=session, objs=metric_instances)
    session.commit()
except:
    session.rollback()
    ...

self._update_latest_metrics_if_necessary(metric_instances, session)
session.commit()
```

Reply:
I discovered an issue when running locally against sqlite where `self._update_latest_metrics_if_necessary(metric_instances, session)` is failing with a duplicate element and the error handling logic in https://github.com/mlflow/mlflow/pull/5460/files#diff-b2b10e69c9c9134afc376963051160539426069d6ed9ebd6af8252ce2b51de09R645-R664 gets stuck. Investigating further...